lib/web_ui/dev/steps/run_tests_step.dart

```diff
@@ -0,0 +1,295 @@
+        testFiles: <FilePath>[failureSmokeTestPath],
+        browserEnvironment: browserEnvironment,
+        concurrency: 1,
+        expectFailure: true,
+        isDebug: isDebug,
+        doUpdateScreenshotGoldens: doUpdateScreenshotGoldens,
+      );
+    }
+
+    // Run all unit-tests as a single batch and with high concurrency.
+    if (unitTestFiles.isNotEmpty) {
+      await _runTestBatch(
+        testFiles: unitTestFiles,
+        browserEnvironment: browserEnvironment,
+        concurrency: 10,
+        expectFailure: false,
+        isDebug: isDebug,
+        doUpdateScreenshotGoldens: doUpdateScreenshotGoldens,
+      );
+      _checkExitCode('Unit tests');
+    }
+
+    // Run screenshot tests one at a time. Otherwise, tests end up
+    // screenshotting each other.
+    for (FilePath screenshotTestFilePath in screenshotTestFiles) {
+      await _runTestBatch(
+        testFiles: <FilePath>[screenshotTestFilePath],
+        browserEnvironment: browserEnvironment,
+        concurrency: 1,
+        expectFailure: false,
+        isDebug: isDebug,
+        doUpdateScreenshotGoldens: doUpdateScreenshotGoldens,
+      );
+      _checkExitCode('Golden tests');
+    }
+
+    if (!allShardsPassed) {
+      throw ToolExit(_createFailedShardsMessage());
+    }
+  }
+
+  void _checkExitCode(String shard) {
+    if (io.exitCode != 0) {
+      failedShards.add(shard);
+    }
+  }
+
+  String _createFailedShardsMessage() {
+    final StringBuffer message = StringBuffer(
+      'The following test shards failed:\n',
+    );
+    for (String failedShard in failedShards) {
+      message.writeln(' - $failedShard');
+    }
+    return message.toString();
+  }
+}
+
+Future<void> _prepareTestResultsDirectory() async {
+  if (environment.webUiTestResultsDirectory.existsSync()) {
+    environment.webUiTestResultsDirectory.deleteSync(recursive: true);
+  }
+  environment.webUiTestResultsDirectory.createSync(recursive: true);
+}
+
+const List<String> _kTestFonts = <String>[
+  'ahem.ttf',
+  'Roboto-Regular.ttf',
+  'NotoNaskhArabic-Regular.ttf',
+  'NotoColorEmoji.ttf',
+];
+
+void _copyTestFontsIntoWebUi() {
+  final String fontsPath = pathlib.join(
+    environment.flutterDirectory.path,
+    'third_party',
+    'txt',
+    'third_party',
+    'fonts',
+  );
+
+  for (String fontFile in _kTestFonts) {
+    final io.File sourceTtf = io.File(pathlib.join(fontsPath, fontFile));
+    final String destinationTtfPath =
+        pathlib.join(environment.webUiRootDir.path, 'lib', 'assets', fontFile);
+    sourceTtf.copySync(destinationTtfPath);
+  }
+}
+
+/// Runs a batch of tests.
+///
+/// Unless [expectFailure] is set to false, sets [io.exitCode] to a non-zero
+/// value if any tests fail.
+Future<void> _runTestBatch({
+  required List<FilePath> testFiles,
+  required bool isDebug,
+  required BrowserEnvironment browserEnvironment,
+  required bool doUpdateScreenshotGoldens,
+  required int concurrency,
+  required bool expectFailure,
```

Comment:
What is the use case for this flag?

Reply:
We run smoke tests that are expected to fail. This is for detecting when there's something wrong and the test suite isn't running at all.